sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveUnion.scala

```diff
@@ -181,7 +100,12 @@ object ResolveUnion extends Rule[LogicalPlan] {
             // like that. We will sort columns in the struct expression to make sure two sides of
             // union have consistent schema.
             aliased += foundAttr
-            Alias(addFields(foundAttr, target), foundAttr.name)()
+            val targetType = try {
+              target.merge(source, conf.resolver)
+            } catch { case e: Throwable =>
+              throw QueryExecutionErrors.failedToMergeIncompatibleSchemasError(target, source, e)
```

Comment:
Should I catch this and re-throw this error or just let the original SparkException be thrown for bad merges?

Reply:
Original exceptions looks more details than just `failedToMergeIncompatibleSchemasError`. I think it is better to let `merge` throw original exceptions.

Reply:
Done